### PR TITLE
[dune] Full Dune support.

### PR DESCRIPTION
**Pull request by ejgallego**

### CODE REVIEW

**Comment by ejgallego** [comment]
How do you usually work in QuickChick wrt the dependencies folks?

I am adding support for depending on Coq libs to Dune, but I wonder how many cases I can cover in the first phase.

**Comment by liyishuai** [comment]
Did you mean libraries "written in Coq" or "provided by Coq"?

**Comment by ejgallego** [comment]
> Did you mean libraries "written in Coq" or "provided by Coq"?

I mean QuickChick dependencies, that is to say, `simple-io`, `mathcomp` etc...

Let's take for example `simple-io`, do you:
- install it, then use it,
- have a local copy, modify the flags

Note that Dune will very soon support composition of Coq libraries in the same way it supports composition of OCaml libs.

**Comment by liyishuai** [comment]
For now we only install them and use their interface, without a local copy.

Also, could you provide an example of how Dune works, so that I could add it to QuickChick CI?

**Comment by ejgallego** [comment]
> For now we only install them and use their interface, without a local copy.

I see, thanks! Note that with Dune you can also use a local copy for example of simple-io and you get a full incremental build of the composition. This is very very useful when hacking in plugins for example, or in certain scenarios where you have control over the base lib.

For example, imagine you want to modify simple-io and quickchick at the same time, with Dune you don't need to do `cd simple-io && make && make install && cd quickchick && make ...` , a `dune build` does understand both libraries as a single unit and will work.

> Also, could you provide an example of how Dune works, so that I could add it to QuickChick CI?

I will take care of this, I suggest we wait until Dune 1.9 is released tho. I'll ping you back for merge then.

**Comment by ejgallego** [comment]
Travis and the Dune build seem to work!

**Comment by liyishuai** [comment]
Does Dune also install the [executable](/ejgallego/QuickChick/blob/ee8e5dc0512c318ab2a50680a3bf8bac4132a1d0/Makefile#L35)?

**Comment by ejgallego** [comment]
> Does Dune also install the [executable](/ejgallego/QuickChick/blob/ee8e5dc0512c318ab2a50680a3bf8bac4132a1d0/Makefile#L35)?

It does now :)

Fixing the tests to properly depend on quickchick does require the composition patch to Dune which should arrive soon. Let's just wait for it instead of doing a hack.

File: examples/multifile-mutation/dune [hunk 1]
```diff
@@ -0,0 +1,2 @@
+(coqlib
+ (name multifile_mutation))
```
Comment (liyishuai) [inline examples/multifile-mutation/dune]: What does this file do? Do I need to add it to all examples?

Comment (ejgallego) [reply examples/multifile-mutation/dune]: This declares the test lib to compile, however it is still not useful as we cannot depend on the quickChick lib itself.

So it is unused at the moment.

**Comment by ejgallego** [comment]
@liyishuai we are installing the tool with the wrong name it seems, I've pushed a fix, but please confirm.

**Comment by ejgallego** [comment]
Oh indeed the problem was different, `.gitignore` was ignoring the dune file for the tool so it wasn't added to the repos at all.

**Comment by ejgallego** [comment]
Ok, so I think this is more or less done for now; support for Coq + Dune is obviously experimental but you can folks merge if you would like. IMHO the Dune workflow for plugins is very interesting.

We will see more news in Dune 1.10.0

**Comment by liyishuai** [comment]
Dune build fails upon warning in later OCaml:
```ocaml
File "src/genericLib.ml", line 141, characters 20-38:
141 |   let compare x y = Pervasives.compare (string_of_qualid x) (string_of_qualid y)
                          ^^^^^^^^^^^^^^^^^^
Error (alert deprecated): module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the 
stdlib-shims library: https://github.com/ocaml/stdlib-shims
```

**Comment by ejgallego** [comment]
> Dune build fails upon warning in later OCaml:

Either disable the warning or fix it.

**Pull request merged**